clpy/core/include/clpy/carray.clh

```diff
@@ -530,54 +532,83 @@ static void __clpy_end_print_out() __attribute__((annotate("clpy_end_print_out")
 
 
 #ifdef __ULTIMA
-__attribute__((annotate("clpy_no_mangle"))) static half convert_float_to_half(float x);
+__attribute__((annotate("clpy_no_mangle"))) static ushort convert_float_to_half_ushort(float x);
+__attribute__((annotate("clpy_no_mangle"))) static float convert_half_ushort_to_float(ushort x);
 #else
 #include "fp16.clh"
+#endif
+
+#ifdef __CLPY_ENABLE_CL_KHR_FP16
+
+#ifndef __ULTIMA
+
 typedef half __clpy__half;
+
 #endif
 
-#ifdef __ULTIMA
-__attribute__((annotate("clpy_no_mangle"))) static half clpy_nextafter_fp16(half x1, half x2);
+__attribute__((annotate("clpy_no_mangle"))) static __clpy__half clpy_nextafter_fp16(__clpy__half x1, __clpy__half x2){
+  return nextafter(x1, x2);
+}
+
+__attribute__((annotate("clpy_no_mangle"))) static __clpy__half convert_float_to_half(float x){
+  const ushort h = convert_float_to_half_ushort(x);
+  return *(const half*)&h;
+}
+
 #else
-static int isnan_fp16(half x){
-  unsigned short const* x_raw = (unsigned short const*)&x;
-  return (*x_raw & 0x7c00u) == 0x7c00u && (*x_raw & 0x03ffu) != 0x0000u;
-}
-static int isfinite_fp16(half x){
-  unsigned short const* x_raw = (unsigned short const*)&x;
-  return (*x_raw & 0x7c00u) != 0x7c00u;
-}
-static int iszero_fp16(half x){
-  unsigned short const* x_raw = (unsigned short const*)&x;
-  return (*x_raw & 0x7fffu) == 0;
-}
-static int eq_nonan_fp16(half x1, half x2){
-  unsigned short const* x1_raw = (unsigned short const*)&x1;
-  unsigned short const* x2_raw = (unsigned short const*)&x2;
-  return (*x1_raw == *x2_raw || ((*x1_raw | *x2_raw) & 0x7fff) == 0);
-}
-static half clpy_nextafter_fp16(half x1, half x2){
-  unsigned short const* x1_raw = (unsigned short const*)&x1;
-  unsigned short const* x2_raw = (unsigned short const*)&x2;
+
+#ifdef __ULTIMA
+static void __clpy_begin_print_out() __attribute__((annotate("clpy_begin_print_out")));
+
+typedef struct __attribute__((packed)) __attribute__((aligned(2))) __clpy__half_{
+  ushort v;
+  __clpy__half_() = default;
+  __clpy__half_(float f):v{convert_float_to_half_ushort(f)}{}
+  operator float()const{return convert_half_ushort_to_float(v);}
```

Comment:
I understand Ultima can solve the problem.

But I don't understand why you need Ultima. As I told you in the some old issue, you tend to solve everything by using Ultima but we must avoid it unless there's no solution without Ultima. Minimize Ultima to minimize maintenance difficulties.

Reply:
Actually, I need the changes in this PR to ultima only for passing `ultima_tests` .
If we decide to change the behavior of `half` identifier, we doesn't need it.
However, we will need it to support all `half` operations, so I proposed this changes.

Reply:
> If we decide to change the behavior of half

As of https://github.com/fixstars/clpy/pull/266#discussion_r376760549

Reply:
> As of #266 (comment), no `half` variable will be constructed by users.

**Until to decide to support all `half` operations** . OK.

Then, when you

> decide to change the behavior of `half` identifier

? Again, **we need to decide to change the behavior of `half` identifier from current ClPy** to disable `cl_khr_fp16` without Ultima. It means that we need to fix or disable some tests. If you have already decided it in your own, you should indicate your opinion here.